BJS2-3492 - Получение пользователей - Unicorn BC3

hhzktoeto opened the pull request.

hhzktoeto committed:
Логика + тесты

Создал эндпоинты
Написал тесты
Протестировал в Postman

hhzktoeto committed:
Решил конфликты

hhzktoeto committed:
Добавления и поправки

Добавил password в UserDto
Добавил PasswordValidationException
Добавил метод create с валидацией пароля
Добавил в UserValidator методы для валидации полей DTO и пароля

hhzktoeto committed:
Поправки + тесты

Добавил тесты на валидацию полей UserDto и валидацию пароля
Мелкие поправки

hhzktoeto commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -30,6 +31,38 @@ public class UserService {
+        return userMapper.toDto(user);
+    }
+
+    public List<UserDto> getUsersByIds(List<Long> ids) {
+        List<User> users = userRepository.findAllById(ids);
+        if (users.isEmpty()) {
+            throw new EntityNotFoundException("Users with given IDs don't exist");
+        }
+        return userMapper.toDto(users);
+    }
```
В случае, если в репозитории не нашлось ни одного юзера с переданными в метод id, возвращается пустой лист. 
Как сделать лучше: 
- пусть пользователю вернётся пустой список
- всё таки выбросить исключение

snailrake replied:
Возвращать пустой список

snailrake replied:
Возвращать пустой список

hhzktoeto commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -1,23 +1,60 @@
-    public UserDto deactivateUser(long userId) {
+    @Operation(summary = "Get user by id")
+    @GetMapping("/users/{userId}")
+    public UserDto getUser(@PathVariable long userId) {
+        if (userId < 1) {
+            throw new DataValidationException("User's id can't be less than 1");
+        }
+        return userService.getUser(userId);
+    }
```
Надо ли здесь проводить валидацию запроса?

snailrake replied:
Моё мнение - в целом можно, чтобы лишний раз в базу не ходить

hhzktoeto committed:
Мелкие поправки

В репозитории на поле Country у юзера стоит nullable = false, для корректной работы метода create добавил поле в Dto

snailrake commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -1,23 +1,60 @@
 
-    public List<UserDto> getPremiumUsers(UserFilterDto filters) {
+    @Operation(summary = "Create new user")
+    @PostMapping("/users/create")
```
Я бы create убрал, по методу и так будет видно, что это создание, а глаголы не особо по правилам rest

snailrake commented on src/main/java/school/faang/user_service/controller/user/UserController.java:
```diff
@@ -1,23 +1,60 @@
 
 import java.util.List;
 
 @RestController
 @RequiredArgsConstructor
+@Tag(name = "Users", description = "Endpoints for managing users")
 public class UserController {
```
Если в каждом эндпоинте используется путь "/users", то можно было бы задать общий путь для всех через @RequestMapping("/users")

hhzktoeto committed:
Исправления после cross-review

IrinaVereshchagina13 commented on src/main/java/school/faang/user_service/exception/PasswordValidationException.java:
```diff
@@ -0,0 +1,8 @@
+    public PasswordValidationException(String message) {
+        super(message);
+    }
+}
```
Это необязательно, мы можем воспользоваться  DataValidationException

IrinaVereshchagina13 commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -30,6 +31,35 @@ public class UserService {
+    private final UserValidator userValidator;
+
+    public UserDto create(UserDto userDto) {
+        userValidator.validatePassword(userDto);
```
Может здесь еще поля дто отвалидируем?

hhzktoeto replied:
Я в контроллере ведь провожу валидацию DTO 🤔
До этого был отдельный метод, сейчас заменил на аннотации

IrinaVereshchagina13 replied:
Справедливо 👍

IrinaVereshchagina13 commented on src/main/java/school/faang/user_service/dto/user/UserDto.java:
```diff
@@ -14,6 +14,8 @@ public class UserDto {
     private String username;
     private String email;
     private String phone;
+    private String password;
     private boolean active;
     private boolean isPremium;
+    private Long countryId;
```
Зачем нам дополнительно поля password и countryId ?

hhzktoeto replied:
У этих полей в entity стоит nullable = false, через postman не мог протестировать метод create без этих полей

hhzktoeto committed:
Merge branch 'unicorn-master-bc3' into feature-BJS2-3492

# Conflicts:
#	src/main/java/school/faang/user_service/controller/user/UserController.java
#	src/main/java/school/faang/user_service/service/user/UserService.java
#	src/test/java/school/faang/user_service/service/user/UserServiceTest.java

hhzktoeto committed:
Исправления после ревью

удалил PasswordValidationException, заменил все использования на DataValidationException
для валидации полей DTO теперь используются аннотации
